packages/0xcert-ethereum-gateway-contracts/src/contracts/actions-gateway.sol

```diff
@@ -403,167 +406,89 @@ contract ActionsGateway is
    * @param _order Data needed for order.
    */
   function _doActionsReplaceZeroAddress(
-    OrderData memory _order
+    OrderData memory _order,
+    address _replaceAddress
   )
     private
   {
     for(uint256 i = 0; i < _order.actions.length; i++)
     {
       require(
-        proxies[_order.actions[i].proxy] != address(0),
+        proxies[_order.actions[i].proxyId].proxyAddress != address(0),
         INVALID_PROXY
       );
 
-      if (_order.actions[i].kind == ActionKind.create)
+      if (proxies[_order.actions[i].proxyId].kind == ActionKind.create)
       {
         require(
-          Abilitable(_order.actions[i].token).isAble(_order.maker, ABILITY_ALLOW_CREATE_ASSET),
+          Abilitable(_order.actions[i].contractAddress).isAble(
+            _order.signers[bytesToUint8(85, _order.actions[i].params)],
+            ABILITY_ALLOW_CREATE_ASSET
+          ),
           SIGNER_NOT_AUTHORIZED
         );
 
-        if (_order.actions[i].to == address(0))
-        {
-          _order.actions[i].to = _order.taker;
+        address to = bytesToAddress(84, _order.actions[i].params);
+        if (to == address(0)) {
+          to = _replaceAddress;
         }
 
-        XcertCreateProxy(proxies[_order.actions[i].proxy]).create(
-          _order.actions[i].token,
-          _order.actions[i].to,
-          _order.actions[i].value,
-          _order.actions[i].param1
+        XcertCreateProxy(proxies[_order.actions[i].proxyId].proxyAddress).create(
+          _order.actions[i].contractAddress,
+          to,
+          bytesToUint256(64, _order.actions[i].params),
+          bytesToBytes32(32, _order.actions[i].params)
         );
       }
-      else if (_order.actions[i].kind == ActionKind.transfer)
+      else if (proxies[_order.actions[i].proxyId].kind == ActionKind.transfer)
       {
-        address from = address(uint160(bytes20(_order.actions[i].param1)));
-
-        if (_order.actions[i].to == address(0))
-        {
-          _order.actions[i].to = _order.taker;
-        }
-
-        if (from == address(0))
-        {
-          from = _order.taker;
+        address from = _order.signers[bytesToUint8(53, _order.actions[i].params)];
+        address to = bytesToAddress(52, _order.actions[i].params);
+        if (to == address(0)) {
```

Comment:
What if the FROM address is the wildcard address?

Maybe the wildcard address wants to pay for the thing they are receiving!

Reply:
Yeah from can be the wildcard address no problem. If the order is dynamic the last signer is always specified as zeroAddress so the its Id can always be used and it does not need to be replaced.

Reply:
@fulldecent comments?

Reply:
Could you please show where the `from` address in the `ActionKind.transfer` is turned into the `_replaceAddress`?

Also I do not see a test on this situation.

Reply:
```
if (_data.signers[signersLength] == address(0))
    {
      // If the last signature is missing then the address we are replacing with is the msg.sender
      // else the address associated with the last signature is used.
      if (signersLength == _signatures.length) {
        anyAddress = msg.sender;
      } else {
        anyAddress = recoverSigner(claim, _signatures[signersLength]);
      }
      _data.signers[signersLength] = anyAddress;
      // If it is not an any taker/signature order and the last signature is missing then the
      // executor (msg.sender) has to be the last signer.
    }
``` 
lines 244 to 254 in actions order.

So if the last address in signers is zeroAddress this means that from is dynamic and in this part we check this and replace that zeroAddress with the actual address. Then the index points to the replacedAddress.

All test that names end with `with any signer` or `with any taker` test this.